feat(client): adds StreamableHttpClientTransport

Aliaksie opened the pull request:
<!-- Provide a brief summary of your changes -->
adds StreamableHttpClientTransport 
## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
WIP

## Breaking Changes
<!-- Will users need to update their code or configurations? -->
No

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [ ] My code follows the repository's style guidelines
- [ ] New and existing tests pass locally
- [ ] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->
WIP

chickenlj commented on mcp/src/main/java/io/modelcontextprotocol/client/transport/StreamableHttpClientTransport.java:
```diff
@@ -0,0 +1,396 @@
+					data = data.substring(0, data.length() - 1);
+				}
+
+				return new FlowSseClient.SseEvent(event, data, id);
```
The right argument sequence here should be `FlowSseClient.SseEvent(id, event, data);`

Aliaksie replied:
Good catch, thanks!

chickenlj commented on mcp/src/main/java/io/modelcontextprotocol/client/transport/StreamableHttpClientTransport.java:
```diff
@@ -0,0 +1,396 @@
+			.connectTimeout(Duration.ofSeconds(10));
+
+		private final HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
+			.header("Accept", "application/json, text/event-stream");
```
I think we should put "Content-type: application/json" request header here. Some servers might not be able to process the request without this header.

chickenlj commented:
Great job @Aliaksie

Here are some of my thoughts after reading this PR and the SDK source code.

I think just adding the McpClientTransport implementation cannot give full support for Streamable HTTP because how it works is inconsistent with HTTP+SSE in many places, which might require some redesign or refactoring of the current SDK implementation:

1. The initialization process is not necessary in Streamable HTTP, it is only required when implementing stateful management. After initialization, all subsequent requests are required to have a  `mcp-session-id header`. However, the current Java SDK implementation enforces checking of initialization status and does not manage the `mcp-session-id` after initialization.
2. `/mcp GET` requests are constrained in the protocol to be used when the client wants to initiate a SSE request proactively, while the current implementation initiates a GET request and establishe SSE session every time on transport connect. Subsequent POST requests also rely on this to send back responses, which can be seen through the operation of the `pendingResponses` field.

Aliaksie commented:
> Great job @Aliaksie
> 
> Here are some of my thoughts after reading this PR and the SDK source code.
> 
> I think just adding the McpClientTransport implementation cannot give full support for Streamable HTTP because how it works is inconsistent with HTTP+SSE in many places, which might require some redesign or refactoring of the current SDK implementation:
> 
> 1. The initialization process is not necessary in Streamable HTTP, it is only required when implementing stateful management. After initialization, all subsequent requests are required to have a  `mcp-session-id header`. However, the current Java SDK implementation enforces checking of initialization status and does not manage the `mcp-session-id` after initialization.
> 2. `/mcp GET` requests are constrained in the protocol to be used when the client wants to initiate a SSE request proactively, while the current implementation initiates a GET request and establishe SSE session every time on transport connect. Subsequent POST requests also rely on this to send back responses, which can be seen through the operation of the `pendingResponses` field.

Hey @chickenlj Thanks a lot for the thoughtful feedback!

1 I added to Streamable HTTP transport code to manage `mcp-session-id`, and I checked documentation (transports and lifecycle ) and it seems to me that initialization request is still required, even in Streamable HTTP  but the server may or may not return a `mcp-session-id`
2 Yes — that’s a really important. I added the logic to make `connect` lazy and optional  with default connect on init for for backward compatibility...

I hope all this makes the SDK compliant with Streamable HTTP's flexibility while still maintaining support for HTTP+SSE fallback

Aliaksie commented:
Hey @tzolov, hope you're doing well! Just wanted to check in — have you had a chance to review this PR? I’d appreciate any feedback when you get a moment. Looking forward to your thoughts!

shaohzhangebay commented:
Hi team, could you please share the timeline to have streamable http for mcp java sdk ?

skyrdis commented:
when this feature could be ready to be parity with typescript and python?
